pkg/checks/diff.go

```diff
@@ -0,0 +1,255 @@
+)
+
+type DiffOptions struct {
+	Client              *http.Client
+	Logger              *log.Logger
+	PackageListFilename string
+	Dir                 string
+	PackagesDir         string
+	ApkIndexURL         string
+	ExistingPackages    map[string]*repository.Package
+}
+
+func NewDiff() *DiffOptions {
+	o := &DiffOptions{
+		Client: http.DefaultClient,
+		Logger: log.New(log.Writer(), "wolfictl check diff: ", log.LstdFlags|log.Lmsgprefix),
+	}
+
+	return o
+}
+
+// Diff compare a newly built apk with the latest in an APK repository, writing teh differences to a file diff.log
+func (o *DiffOptions) Diff() error {
+	// create two temp folders we can use to explode the apks and compare their contents
+	dirExistingApk, err := os.MkdirTemp("", "wolfictl-apk-*")
+	if err != nil {
+		return errors.Wrapf(err, "failed to create temporary dir")
+	}
+	defer os.RemoveAll(dirExistingApk)
+
+	dirNewApk, err := os.MkdirTemp("", "wolfictl-apk-*")
+	if err != nil {
+		return errors.Wrapf(err, "failed to create temporary dir")
+	}
+	defer os.RemoveAll(dirNewApk)
+
+	// get the latest APKINDEX
+	apkContext := apk.New(o.Client, o.ApkIndexURL)
+	o.ExistingPackages, err = apkContext.GetApkPackages()
+	if err != nil {
+		return errors.Wrapf(err, "failed to get APK packages from URL %s", o.ApkIndexURL)
+	}
+
+	// get a list of new package names that have recently been built
+	newPackages, err := getNewPackages(o.PackageListFilename)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get new packages from file %s", o.PackageListFilename)
+	}
+
+	// for each new package being built grab the latest existing one
+	for newPackageName, newAPK := range newPackages {
+		o.Logger.Printf("checking %s", newPackageName)
+		// read new apk
+		filename := filepath.Join(o.PackagesDir, newAPK.Arch, fmt.Sprintf("%s-%s-r%s.apk", newPackageName, newAPK.Version, newAPK.Epoch))
+		newFile, err := os.Open(filename)
+		if err != nil {
+			return errors.Wrapf(err, "failed to read %s", filename)
+		}
+
+		err = tar.Untar(newFile, filepath.Join(dirNewApk, newPackageName))
+		if err != nil {
+			return errors.Wrapf(err, "failed to untar new apk")
+		}
+
+		// fetch current latest apk
+		p, ok := o.ExistingPackages[newPackageName]
+		if !ok {
+			return fmt.Errorf("no package %s in APKINDEX", newPackageName)
+		}
+
+		existingFilename := fmt.Sprintf("%s-%s.apk", p.Name, p.Version)
+		err = downloadCurrentAPK(o.Client, o.ApkIndexURL, existingFilename, filepath.Join(dirExistingApk, newPackageName))
+		if err != nil {
+			return errors.Wrapf(err, "failed to download %s using base URL %s", newPackageName, existingFilename)
+		}
+	}
+
+	rs, err := diffDirectories(dirNewApk, dirExistingApk, newPackages)
+	if err != nil {
+		return err
+	}
+
+	diffFile := filepath.Join(o.Dir, "diff.log")
+	err = writeDiffLog(rs, diffFile, newPackages)
+	if err != nil {
+		return errors.Wrap(err, "failed writing to file")
+	}
+
+	fmt.Printf("diff written to %s\n", diffFile)
+	return nil
```

Comment:
I think `check ...` commands should probably exit non-zero if a thing being checked is not what's expected. If it's not that kind of command, perhaps it should go somewhere else in the command hierarchy than in `check`.

Reply:
not sure I follow this one, the function does returns errors which fails the check if there's unexpected behaviour.  Unless I'm miss-understanding the comment here?  If I am and it's an easy change mind suggesting an edit so I can follow?

Reply:
the errors are also returned here https://github.com/wolfi-dev/wolfictl/pull/167/files#diff-12d09c941117ea670506fe1b328cbbdd29de8f5eb50bab939cde1febeaaf78e4R20

Reply:
I might have this wrong, but my understanding was that the `check` commands are kind of like linting, in that they are designed to be helpful as CI checks, and you can rely on them to exit non-zero if the thing being "checked" was not true.

I'm trying to map this understanding onto the `wolfictl check diff` command. What are we checking for? What means the check passed, and what means the check failed? Does the fact that there's a diff mean our expectation wasn't met? This seems to be true for the other `wolfictl check ...` commands: `update` and `soname`.

And if we're not asserting a thing here, I'm wondering if it'd be a better organization of CLI commands (and easier to intuit) to move the `diff` command somewhere else, like maybe directly under `wolfictl`?

Reply:
I put it under checks as it shares common code with other checks like soname.  The check diff is more of a utility to help maintainers check differences rather than a failed lint.  Personally I think that's ok but if it's really a blocker I can refactor.

Reply:
Not a blocker! Let's move this to the back of our minds as we're thinking about command organization.